cmd/core/main.go

```diff
@@ -71,38 +104,138 @@ func main() {
+		setupLog.Error(err, "unable to add uploads http handler to manager")
+		os.Exit(1)
+	}
+	if err := mgr.Add(uploadmgr.NewBundleGC(mgr.GetCache(), uploadStorageDirectory, uploadStorageSyncInterval)); err != nil {
+		setupLog.Error(err, "unable to add bundle garbage collector to manager")
+		os.Exit(1)
+	}
+
+	// This finalizer logic MUST be co-located with this main
+	// controller logic because it deals with cleaning up bundle data
+	// from the bundle cache when the bundles are deleted. The
+	// consequence is that this process MUST remain running in order
+	// to process DELETE events for bundles that include this finalizer.
+	// If this process is NOT running, deletion of such bundles will
+	// hang until $something removes the finalizer.
+	//
+	// If the bundle cache is backed by a storage implementation that allows
+	// multiple writers from different processes (e.g. a ReadWriteMany volume or
+	// an S3 bucket), we could have separate processes for finalizer handling
+	// and the primary provisioner controllers. For now, the assumption is
+	// that we are not using such an implementation.
+	bundleFinalizers := crfinalizer.NewFinalizers()
+	if err := bundleFinalizers.Register(finalizer.DeleteCachedBundleKey, &finalizer.DeleteCachedBundle{Storage: bundleStorage}); err != nil {
+		setupLog.Error(err, "unable to register finalizer", "finalizerKey", finalizer.DeleteCachedBundleKey)
+		os.Exit(1)
+	}
+
+	unpacker, err := source.NewDefaultUnpacker(mgr, ns, unpackImage, baseUploadManagerURL, rootCAs)
+	if err != nil {
+		setupLog.Error(err, "unable to setup bundle unpacker")
+		os.Exit(1)
+	}
+
+	if err = (&plaincontrollers.BundleReconciler{
+		Client:     mgr.GetClient(),
+		Scheme:     mgr.GetScheme(),
+		Storage:    bundleStorage,
+		Finalizers: bundleFinalizers,
+		Unpacker:   unpacker,
+	}).SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", rukpakv1alpha1.BundleKind)
+		os.Exit(1)
+	}
+
+	if err = (&registrycontrollers.BundleReconciler{
+		Client:     mgr.GetClient(),
+		Scheme:     mgr.GetScheme(),
+		Storage:    bundleStorage,
+		Finalizers: bundleFinalizers,
+		Unpacker:   unpacker,
+	}).SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", rukpakv1alpha1.BundleKind)
+		os.Exit(1)
+	}
+
+	cfgGetter := helmclient.NewActionConfigGetter(mgr.GetConfig(), mgr.GetRESTMapper(), mgr.GetLogger())
+	if err = (&plaincontrollers.BundleDeploymentReconciler{
+		Client:             mgr.GetClient(),
+		Scheme:             mgr.GetScheme(),
+		BundleStorage:      bundleStorage,
+		ReleaseNamespace:   ns,
+		ActionClientGetter: helmclient.NewActionClientGetter(cfgGetter),
+	}).SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", rukpakv1alpha1.BundleDeploymentKind)
 		os.Exit(1)
 	}
```

Comment:
Perhaps we could do this in an iterative way so that as we add more provisioners this doesn't scale too hard? Could be a follow-up so it isn't blocking.

Reply:
Let's consider for a follow-up, if that's alright. Looping or extracting into a separate function seems like it could move toward the direction of the `NewUnpacker` signature, which we've already raised as a potential rough edge because of all the required parameters.